compiler: define command and implement a simple compiler (#128)
tomarrell/lbadd: doc/intermediary-representation.md

# Commands

> Given that we parser SQL input into an AST, we somehow have to process that produced AST.

Author: ```suggestion
Given that we parse SQL input into an AST, we somehow have to process that produced AST.
```

The finished IR is close to relational algebra, however it is a type structure rather than a string or unreadable byte code.
For now, the IR is not fully defined.
However, basic functionality exists.
A quick note on the IR documentation: Whenever it says that a struct "does anything", it doesn't do it, it just is a command for the executor to do that thing.
For example, if the documentation on `Scan` says "scans a table", what it really means is, that if the executor encounters that instruction, it has to scan that table.

### Basic operations

The IR is a set of commands that can instruct our executor.
This is because "IR" and "commands" are synonyms in this document.

Commands can be parameterized and configured.
Configuration is written in brackets, e.g. `Command[config1=value1,config2=value2]()`.
Parameters are written in parenthesis, e.g. `Command[](param1,param2)`.
Parameters are always evaluated before the command, configuration is evaluated while the command is run.

The configuration and parameters are just separated for sake of readability.
They are fields in the same struct, somewhere in our `command` package.
configuration generally influences the parameterization.

> As an example, see the `Project` command.

Author: Don't know if it'll stay in `internal/compiler/command` or if it will move to `internal/command` or even `internal/ir`

The input list is a list of datasets, but the columns configuration influences, which part of the input is considered.

#### `Scan`
The most basic command is `Scan`.
It scans a table and returns all datasets in that table.
Notice the previous sentence actually means: "It tells the executor to consider all datasets in that table".

> Please be aware that `Scan` does not actually loads all datasets from the indicated table, it just tells the executor to consider them all.

Author: Then I'll have to clarify:
`Scan` is not defined to cause a whole table-scan, and does not immediately imply, that all datasets from the table are loaded into memory.

> Please be aware that `Scan` does not actually loads all datasets from the indicated table, it just tells the executor to consider them all.

Author: ```suggestion
Please be aware that `Scan` does not actually loads all datasets from the indicated table **into memory**, it just tells the executor to consider them all.
```

| Severus | 38 |
| Sam | 22 |

</details>

#### `Select`
Another very basic command is `Select`, which is the equivalent to the relationally algebraic "select" (&sigma;).
`Select` is configured with a filter, which is the subscript to the relational algebra equivalent, and parameterized with the input.
It returns a table with datasets that all satisfy the filter expression.
This implies, that the expression must be able to be evaluated to a boolean value.

| Relational algebra | Command |
|---|---|
| &sigma;<sub>age&ge;25</sub>(Person) | <code>Select[filter=age&ge;25]\(Scan[table=Person]\(\)\)</code> |

In this case, the expression is valid, since age&ge;25 can be evaluated to a boolean value.
An invalid expression would be `"twelve"`.
However, <code>1&ne;2</code> is valid, though it is constant and may be optimized away.

<details>
<summary><code>Select[filter=age&ge;25](Scan[table=Person]())</code></summary>

| Name | Age |
|---|---|
| Sandra | 43 |
| Elsa | 65 |
| Serious | 36 |
| Severus | 38 |

</details>

#### `Project`
The third basic command is `Project`.
The relationally algebraic equivalent also is "project" (&Pi;)
Simply put, it limits what columns are considered.

> The `Project` command is parameterized with a list of input datasets and returns the datasets in the same orderas it received them, except that it removed all columns that should not be considered.

Author: ```suggestion
The `Project` command is parameterized with a list of input datasets and returns the datasets in the same order as it received them, except that it removed all columns that should not be considered.
```

> For example, the following expression yields nothing.
> 
> `Project` is configured with a list of columns that the input is filtered with.
> This implies, that the expressions must be able to be evaluated to literal values.
> 
> | Relational algebra | Command |
> |---|---|
> | &Pi;<sub>Name,Age</sub>(Person) | <code>Project[cols=Name,Age]\(Scan[table=Person]\(\)\)</code> |
> 
> With `Project`, a few opportunities for optimization open up, for example two projections that cancel each other out.
> This is, where the `Empty` command comes into play.
> More on that command is further down below.
> In the following example, first, the inner projection removes all columns that are not `Age`, then, the outer projection removes all columns that are not `Name`, yielding an empty list.
> 
> &Pi;<sub>Name</sub>(&Pi;<sub>Age</sub>(Person))

Author: ```suggestion

`Project` is configured with a list of columns that the input is filtered with.
This implies, that the expressions must be able to be evaluated to literal values.

| Relational algebra | Command |
|---|---|
| &Pi;<sub>Name,Age</sub>(Person) | <code>Project[cols=Name,Age]\(Scan[table=Person]\(\)\)</code> |

With `Project`, a few opportunities for optimization open up, for example two projections that cancel each other out.
This is, where the `Empty` command comes into play.
More on that command is further down below.
In the following example, first, the inner projection removes all columns that are not `Age`, then, the outer projection removes all columns that are not `Name`, yielding an empty list.

&Pi;<sub>Name</sub>(&Pi;<sub>Age</sub>(Person))
```
